### ISSUE TITLE
[Epic] Introduce logical client/server separation within the CLI architecture 

### ISSUE THREAD

**Issue by jaideepr97** [instructlab/instructlab#2325]
**What is the current behavior of feature you are looking to change?**
Currently, the instructLab CLI is implemented in a tightly coupled manner between the user-facing parts of it and the underlying feature implementations. This is mainly done via user-facing click functions flowing directly into the function implementations in a consolidated manner. 
Downsides of this architecture are:
- It makes it hard to test partial code paths if you just want to test either the user-facing side of the flow, or just the actual implementation of the function. 
- It is also incompatible with an API layer that we may soon want to introduce and have the CLI interface with.

**What is the proposed behavior of feature you are looking to change?**
Introduction of logical separation in the CLI architecture, such that we end up with a client/server split in the architecture of the codebase. This can be achieved by introducing a new layer of "handler" functions in between the click end-points, and the functions that contain the actual implementations of various commands and features. 

This will also require the introduction of an interface, that can be used to delegate work either to the embedded handler functions vs a potential remote end point 

**What is the benefit of the design change you are proposing?**
- Splits up the codebase into a easily recognizable client/server pattern 
- introduction of "handler" functions acts as a mock API layer. This allows us to mimic what a potential API should look like by defining the function signatures for these handler functions to fit our needs
- This is a low-risk change, with no breakage to existing CLI users
- Changing the function signatures is way easier than changing a fixed and agreed upon API spec that may have many more consumers outside of this project. This sets us up for experimentation before we actually design an API layer 
- Loose coupling between 'client' and 'server' side allow for better mocking and testing of individual commands and functional endpoints

**If applicable, how would backwards-compatability be handled?**
No breaking changes to worry about 

**Additional context**
This work sets us up to design a REST API that can eventually be consumed by the CLI, alongside other stakeholders 

Related sub-tasks:
- [x] #2326 
- [x] #2328 
- [x] #2329 
- [x] #2331 
- [x] #2332 
- [x] #1548

**Comment by cdoern** [comment]
@jaideepr97 should we split this into different issues for implementing each part of this handler setup? Like one for all model cmds?

**Comment by jaideepr97** [comment]
> @jaideepr97 should we split this into different issues for implementing each part of this handler setup? Like one for all model cmds?

@cdoern yeah that's a good idea, this one can be converted into an epic and sub-issues can be linked here - split by cmd group
